### ISSUE TITLE
Instructions for running `kube-bench` on AWS EKS

### ISSUE THREAD

**Issue by complex64** [aquasecurity/kube-bench#220]
We are looking to run `kube-bench` against an EKS cluster in AWS. This is EKS platform `eks.1` at Kubernetes version `1.11`, i.e. everything is up-to-date.

We have been taking the approach of running the job YAMLs included in this repository. However with no success. I'll now try to document the issues as well as possible, and would be grateful for your help in resolving this.

--- 

## 1. The EKS control plane is not user accessible

EKS, as a managed service offered by AWS, does not permit access to the control plane, that is master nodes, to schedule workloads.

The node selecor and tolerations in [job-master.yaml](https://raw.githubusercontent.com/aquasecurity/kube-bench/master/job-master.yaml) have no effect. The pod will not be scheduled anywhere and the job's pod will be left in state "Pending" indefinitely. Example pod events for a pod `kube-bench-master-wlmxl`:

```
  Type     Reason             Age                 From                Message
  ----     ------             ----                ----                -------
  Warning  FailedScheduling   11s (x25 over 35s)  default-scheduler   0/9 nodes are available: 9 node(s) didn't match node selector.
```
What approach would you recommend instead for security benchmarks of AWS EKS master nodes?

## 2. The Node Job has $PATH issues on AWS EKS

We are adjusting the command in [job-node.yaml](https://raw.githubusercontent.com/aquasecurity/kube-bench/master/job-node.yaml) for `-v3` prior running.

The job's pods will cycle quickly from Pending, ContainerCreating, to Error. Inspection of the logs yields:

```
I0219 11:26:28.944088   20794 util.go:132] Looking for config for version 1.11
I0219 11:26:28.944134   20794 util.go:137] Looking for config file: cfg/1.11/node.yaml
I0219 11:26:28.944175   20794 common.go:65] Using benchmark file: cfg/1.11/node.yaml
I0219 11:26:28.944408   20794 common.go:77] Using config file: cfg/1.11/config.yaml
I0219 11:26:28.944632   20794 util.go:57] [ps -C hyperkube -o cmd --no-headers]: fork/exec /usr/bin/ps: no such file or directory
I0219 11:26:28.944662   20794 util.go:284] executable 'hyperkube kubelet' not running
I0219 11:26:28.944819   20794 util.go:57] [ps -C kubelet -o cmd --no-headers]: fork/exec /usr/bin/ps: no such file or directory
I0219 11:26:28.944844   20794 util.go:284] executable 'kubelet' not running

need kubelet executable but none of the candidates are running
```

I'd expect `ps` to be in `/bin`, so this looked like a `$PATH` issue. Handing the job a PATH through the environment:

```yaml
        env:
          - name: PATH
            value: /bin:/usr/bin:/usr/local/bin
```

Will in fact allow the job to run to completion.

Perhaps a change in documentation or job YAML is sufficient here, what do you think?

**Comment by lizrice** [comment]
There is also a significant issue as described [here](https://github.com/aquasecurity/kube-bench/issues/195) - EKS uses a newer method of making config settings which isn't supported by kube-bench yet (though we plan to address that soon). 

For 1, I am not sure whether it's possible to schedule onto EKS master nodes, I will look into that. 

For 2, that change in env var looks fine to me

**Comment by complex64** [comment]
Thanks for looking into this, much appreciated!

I'll be filing a PR then for the env-var change.

**Comment by Constantin07** [comment]
I'm also having issues running it on EKS worker nodes.
```bash
kubectl run --rm -i -t kube-bench-node --image=aquasec/kube-bench:latest --restart=Never --overrides="{ \"apiVersion\": \"v1\", \"spec\": { \"hostPID\": true } }" -- node --version 1.11
```

I'm getting output:
```bash
[INFO] 2 Worker Node Security Configuration
[INFO] 2.1 Kubelet
[FAIL] 2.1.1 Ensure that the --allow-privileged argument is set to false (Scored)
[PASS] 2.1.2 Ensure that the --anonymous-auth argument is set to false (Scored)
[PASS] 2.1.3 Ensure that the --authorization-mode argument is not set to AlwaysAllow (Scored)
[FAIL] 2.1.4 Ensure that the --client-ca-file argument is set as appropriate (Scored)
[PASS] 2.1.5 Ensure that the --read-only-port argument is set to 0 (Scored)
[PASS] 2.1.6 Ensure that the --streaming-connection-idle-timeout argument is not set to 0 (Scored)
[FAIL] 2.1.7 Ensure that the --protect-kernel-defaults argument is set to true (Scored)
[PASS] 2.1.8 Ensure that the --make-iptables-util-chains argument is set to true (Scored)
[PASS] 2.1.9 Ensure that the --hostname-override argument is not set (Scored)
[PASS] 2.1.10 Ensure that the --event-qps argument is set to 0 (Scored)
[FAIL] 2.1.11 Ensure that the --tls-cert-file and --tls-private-key-file arguments are set as appropriate (Scored)
[PASS] 2.1.12 Ensure that the --cadvisor-port argument is set to 0 (Scored)
[FAIL] 2.1.13 Ensure that the --rotate-certificates argument is not set to false (Scored)
[PASS] 2.1.14 Ensure that the RotateKubeletServerCertificate argument is set to true (Scored)
[WARN] 2.1.15 Ensure that the Kubelet only makes use of Strong Cryptographic Ciphers (Not Scored)
[INFO] 2.2 Configuration Files
[FAIL] 2.2.1 Ensure that the kubelet.conf file permissions are set to 644 or more restrictive (Scored)
[FAIL] 2.2.2 Ensure that the kubelet.conf file ownership is set to root:root (Scored)
[FAIL] 2.2.3 Ensure that the kubelet service file permissions are set to 644 or more restrictive (Scored)
[FAIL] 2.2.4 Ensure that the kubelet service file ownership is set to root:root (Scored)
[FAIL] 2.2.5 Ensure that the proxy kubeconfig file permissions are set to 644 or more restrictive (Scored)
[FAIL] 2.2.6 Ensure that the proxy kubeconfig file ownership is set to root:root (Scored)
[WARN] 2.2.7 Ensure that the certificate authorities file permissions are set to 644 or more restrictive (Scored)
[WARN] 2.2.8 Ensure that the client certificate authorities file ownership is set to root:root (Scored)
[FAIL] 2.2.9 Ensure that the kubelet configuration file ownership is set to root:root (Scored)
[FAIL] 2.2.10 Ensure that the kubelet configuration file has permissions set to 644 or more restrictive (Scored)
```

The section 2.2 Configuration Files mistakenly reports permissions for the files which actually do not exist on EKS worker node (used the latest Amazon EKS Optimised image).
In my opinion- if the file/dir doesn't exist - it should not misleadingly report a `FAIL`.

**Comment by lizrice** [comment]
@Constantin07 is it possible that the files exist, but are in different locations or have different filenames in EKS?  Options for the filenames and paths can be configured through the cfg/config.yaml file - kube-bench will check through all the listed alternatives until it finds one. We would happily take a PR to add alternative locations as required for additional platforms like EKS.

**Comment by lizrice** [comment]
This should now be addressed through #260
